src/schema/property.ts

```diff
@@ -1,69 +1,19 @@
-import { DecoratorType } from '../enums';
-import {
-    IsBooleanOptions,
-    IsDateOptions,
-    IsDateStringOptions,
-    IsEnumOptions,
-    IsIntegerOptions,
-    IsNestedOptions,
-    IsNumberOptions,
-    IsStringOptions,
-    IsUUIDOptions,
-} from '../options';
+import { DTODecoratorName, DTODecoratorOptions } from '../types';
 
-export interface IsBooleanProperty {
-    decorator: DecoratorType.IsBoolean,
-    options: IsBooleanOptions,
-}
-
-export interface IsDateProperty {
-    decorator: DecoratorType.IsDate,
-    options: IsDateOptions,
-}
-
-export interface IsDateStringProperty {
-    decorator: DecoratorType.IsDateString,
-    options: IsDateStringOptions,
-}
-
-export interface IsEnumProperty {
-    decorator: DecoratorType.IsEnum,
-    options: IsEnumOptions,
-}
-
-export interface IsIntegerProperty {
-    decorator: DecoratorType.IsInteger,
-    options: IsIntegerOptions,
-}
-
-export interface IsNestedProperty {
-    decorator: DecoratorType.IsNested,
-    options: IsNestedOptions,
-}
-
-export interface IsNumberProperty {
-    decorator: DecoratorType.IsNumber,
-    options: IsNumberOptions,
-}
-
-export interface IsStringProperty {
-    decorator: DecoratorType.IsString,
-    options: IsStringOptions,
-}
-
-export interface IsUUIDProperty {
-    decorator: DecoratorType.IsUUID,
-    options: IsUUIDOptions,
+export interface IsSchemaProperty<Name extends DTODecoratorName> {
+    decorator: Name,
+    options: DTODecoratorOptions<Name>,
 }
 
 /* eslint-disable @typescript-eslint/indent */
-export type SchemaProperty = IsBooleanProperty
-    | IsDateProperty
-    | IsDateStringProperty
-    | IsEnumProperty
-    | IsIntegerProperty
-    | IsNestedProperty
-    | IsNumberProperty
-    | IsStringProperty
-    | IsUUIDProperty;
+export type SchemaProperty =
```

Comment:
If we don't define this as a union, `tsc` has a hard time understanding which combinations of names and options are legal in our `Schema` type (next file).